src/bit-systems/audio-zone-system.ts

```diff
@@ -0,0 +1,318 @@
+    const inOutParams = Array.from(sourceZones)
+      .filter(zone => AudioZone.flags[zone] & AUDIO_ZONE_FLAGS.IN_OUT)
+      .filter(exclude(listenerZones))
+      .filter(hasIntersection(ray))
+      .map(zone => getZoneAudioParams(zone))
+      .reduce(paramsReducer, null);
+
+    // TODO: Reimplement the desired sorting of zones
+    const outInParams = Array.from(listenerZones)
+      .filter(zone => AudioZone.flags[zone] & AUDIO_ZONE_FLAGS.OUT_IN)
+      .filter(exclude(sourceZones))
+      .filter(hasIntersection(ray))
+      .map(zone => getZoneAudioParams(zone))
+      .reduce(paramsReducer, null);
+
+    if (!outInParams && !inOutParams) {
+      restoreSourceParams(source);
+    } else if (outInParams && !inOutParams) {
+      applySourceParams(source, outInParams);
+    } else if (!outInParams && inOutParams) {
+      applySourceParams(source, inOutParams);
+    } else {
+      // In this case two zones ar acting over the same source simultaneously.
+      // We apply the closest zone params with the lowest gain
+      applySourceParams(
+        source,
+        Object.assign(
+          {},
+          inOutParams,
+          outInParams,
+          paramsReducer(
+            {
+              gain: outInParams?.gain
+            },
+            {
+              gain: inOutParams?.gain
+            }
+          )
+        )
+      );
+    }
+  };
+})();
+
+const REDUCED_KEYS = [
+  "gain",
+  "maxDistance",
+  "refDistance",
+  "rolloffFactor",
+  "coneInnerAngle",
+  "coneOuterAngle",
+  "coneOuterGain"
+] as const;
+type ReducedAudioSettingsKeys = typeof REDUCED_KEYS[number];
+type ReducedAudioSettings = Pick<AudioSettings, ReducedAudioSettingsKeys>;
+
+// We apply the most restrictive audio parameters
+const paramsReducer = (acc: AudioSettings, curr: AudioSettings): AudioSettings => {
+  if (!curr && !acc) return {} as AudioSettings;
+  else if (curr && !acc) return curr;
+  else if (!curr && acc) return acc;
+  else
+    return REDUCED_KEYS.reduce((result: ReducedAudioSettings, key: ReducedAudioSettingsKeys): AudioSettings => {
+      if (curr[key] !== undefined && acc[key] !== undefined) {
+        result[key] = Math.min(acc[key]!, curr[key]!);
+      } else if (curr[key] !== undefined && acc[key] === undefined) {
+        result[key] = curr[key];
+      } else if (curr[key] === undefined && acc[key] !== undefined) {
+        result[key] = acc[key];
+      }
+      return result;
+    }, {});
+};
+
+const addOrRemoveZone = (zones: Set<number>, zone: number, position: Vector3) => {
+  if (!zones) return;
+  const isInZone = isZoneEnabled(zone) && zoneContains(zone, position);
+  const wasInZone = zones.has(zone);
+  if (isInZone && !wasInZone) {
+    zones.add(zone);
+  } else if (!isInZone && wasInZone) {
+    zones.delete(zone);
+  }
+};
+
+const currZones = new Map<number, Set<number>>();
+const prevZones = new Map<number, Set<number>>();
+const aabbs = new Map<number, Box3>();
```

Comment:
It seems like this is (non-numeric) component data that we can store using the standard "escape hatch".

Reply:
I would like to know more about what's the benefit of following the escape hatch pattern vs having module level maps. If I was an external client developer what would following this pattern buy me in this case? At first glance the benefit is having a entry point for all component bound data structures and I think that's a good thing. In this case this maps are storing transient data that don't have much interest for external systems. Is this assumption right or am I missing the point?